### PR TITLE
feat: add C implementation for `math/base/special/modf`

### LINKED ISSUE #1916: [RFC]: add C implementation for `math/base/special/modf`
### Description

This RFC proposes adding C implementation for [`math/base/special/modf`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/modf).

### Related Issues

Related issues #649.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by gunjjoshi**
Resolves #1916.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/modf`](https://github.com/stdlib-js/stdlib/tree/d7798af1c843b45d0bbaddadedf7d2961a1b3f57/lib/node_modules/%40stdlib/math/base/special/modf)

```C
void stdlib_base_modf( const double x, double* integral, double* frac )
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649.
-   fixes #1916.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   ☑️ Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/modf/README.md [hunk 1]
```diff
@@ -105,6 +105,95 @@ for ( i = 0; i < 100; i++ ) {
...
+```
+
+#### stdlib_base_modf( x, integral, frac )
+
+Decomposes a [double-precision floating-point number][ieee754] into integral and fractional parts, each having the same type and sign as `x`.
+
+```c
+double integral;
+double frac;
+stdlib_base_modf( 4.0, &integral, &frac );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/README.md]:
```suggestion
double frac;

stdlib_base_modf( 4.0, &integral, &frac );
```

File: lib/node_modules/@stdlib/math/base/special/modf/README.md [hunk 2]
```diff
@@ -105,6 +105,95 @@ for ( i = 0; i < 100; i++ ) {
...
+
+int main( void ) {
+    const double x[] = { 4.0, 0.0, -0.0, 1.0, -1.0, 3.14, -3.14, 1.0e308, -1.0e308, 1.0/0.0, -1.0/0.0, 0.0/0.0 };
+
+    double integral;
+    double frac;
+    int i;
+    for ( i = 0; i < 12; i++ ) {
+        stdlib_base_modf( x[i], &integral, &frac );
+        printf( "x: %f => integral: %f, frac: %f\n", x[i], integral, frac );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/README.md]:
```suggestion
        printf( "x: %lf => integral: %lf, frac: %lf\n", x[ i ], integral, frac );
```

File: lib/node_modules/@stdlib/math/base/special/modf/README.md (hunk 1 above)
Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/modf/README.md]: Please do not commit this atm

File: lib/node_modules/@stdlib/math/base/special/modf/README.md [hunk 3]
```diff
@@ -105,6 +105,95 @@ for ( i = 0; i < 100; i++ ) {
...
+#include <stdio.h>
+
+int main( void ) {
+    const double x[] = { 4.0, 0.0, -0.0, 1.0, -1.0, 3.14, -3.14, 1.0e308, -1.0e308, 1.0/0.0, -1.0/0.0, 0.0/0.0 };
+
+    double integral;
+    double frac;
+    int i;
+    for ( i = 0; i < 12; i++ ) {
+        stdlib_base_modf( x[i], &integral, &frac );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/README.md]:
```suggestion
        stdlib_base_modf( x[ i ], &integral, &frac );
```

File: lib/node_modules/@stdlib/math/base/special/modf/benchmark/c/native/benchmark.c [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+/**
+* Runs a benchmark.
+*
+* @return elapsed time in seconds
+*/
+double benchmark() {
+	double elapsed;
+	double integral;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/benchmark/c/native/benchmark.c]:
```suggestion
	double integral;
	double elapsed;
```

File: lib/node_modules/@stdlib/math/base/special/modf/examples/c/example.c [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+int main( void ) {
+    const double x[] = { 4.0, 0.0, -0.0, 1.0, -1.0, 3.14, -3.14, 1.0e308, -1.0e308, 1.0/0.0, -1.0/0.0, 0.0/0.0 };
+
+    double integral;
+    double frac;
+    int i;
+    for ( i = 0; i < 12; i++ ) {
+        stdlib_base_modf( x[i], &integral, &frac );
+        printf( "x: %f => integral: %f, frac: %f\n", x[i], integral, frac );
+    }
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/examples/c/example.c]:
```suggestion
	const double x[] = { 4.0, 0.0, -0.0, 1.0, -1.0, 3.14, -3.14, 1.0e308, -1.0e308, 1.0/0.0, -1.0/0.0, 0.0/0.0 };

	double integral;
	double frac;
	int i;
	for ( i = 0; i < 12; i++ ) {
		stdlib_base_modf( x[ i ], &integral, &frac );
		printf( "x: %lf => integral: %lf, frac: %lf\n", x[ i ], integral, frac );
	}
```

File: lib/node_modules/@stdlib/math/base/special/modf/manifest.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-infinite",
+        "@stdlib/math/base/assert/is-nan",
+        "@stdlib/number/float64/base/from-words",
+        "@stdlib/number/float32/base/to-word",
+        "@stdlib/constants/float64/pinf",
+        "@stdlib/constants/float64/exponent-bias",
+        "@stdlib/constants/float64/high-word-exponent-mask",
+        "@stdlib/constants/float64/high-word-significand-mask"
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/manifest.json]:
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/number/float64/base/from-words",
        "@stdlib/number/float64/base/to-words",
        "@stdlib/constants/float64/pinf",
        "@stdlib/constants/float64/exponent-bias",
        "@stdlib/constants/float64/high-word-exponent-mask",
        "@stdlib/constants/float64/high-word-significand-mask"
```

Also, this is not the latest type of manifest that we use, please check

File: lib/node_modules/@stdlib/math/base/special/modf/src/addon.c [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	status = napi_get_value_double( env, argv[ 1 ], &value );
+	assert( status == napi_ok );
+
+	double integral;
+	double frac;
+	stdlib_base_modf( value, &integral, &frac );
+
+	double *op = (double *)Out;
+	op[ 1 ] = integral;
+	op[ 0 ] = frac;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/addon.c]:
```suggestion
	op[ 0 ] = integral;
	op[ 1 ] = frac;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+#include "stdlib/number/float64/base/from_words.h"
+#include "stdlib/number/float64/base/to_words.h"
+#include "stdlib/constants/float64/pinf.h"
+#include "stdlib/constants/float64/exponent_bias.h"
+#include "stdlib/constants/float64/high_word_exponent_mask.h"
+#include "stdlib/constants/float64/high_word_significand_mask.h"
+
+
+// VARIABLES //
+
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
```

We do not have these headers in C

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		return;
+	}
+	if ( stdlib_base_is_nan( x ) ) {
+		*integral = 0.0 / 0.0; // NaN
+		*frac = 0.0 / 0.0; // NaN
+		return;
+	}
+	if ( x == STDLIB_CONSTANT_FLOAT64_PINF ) {
+		*frac = STDLIB_CONSTANT_FLOAT64_PINF;
+		*integral = 0.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
		*integral = STDLIB_CONSTANT_FLOAT64_PINF;
		*frac = 0.0;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+	exp -= ( STDLIB_CONSTANT_FLOAT64_EXPONENT_BIAS );
+
+	// Handle smaller values (x < 2**20 = 1048576)...
+	if( exp < 20 ) {
+		i = ( STDLIB_CONSTANT_FLOAT64_HIGH_WORD_SIGNIFICAND_MASK >> exp );
+
+		// Determine if `x` is integral by checking for significand bits which cannot be exponentiated away...
+		if ( ( ( high & i ) | low ) == 0 ) {
+			*frac = x;
+			*integral = 0.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
			*integral = x;
			*frac = 0.0;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+			return;
+		}
+		// Turn off all the bits which cannot be exponentiated away:
+		high &= ( ~i );
+
+		// Generate the integral part:
+		j = i;
+		stdlib_base_float64_from_words( high, 0, &j );
+		*frac = j;
+		*integral = x - j;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
		stdlib_base_float64_from_words( high, 0, &j );
		
		// The fractional part is whatever is leftover:
		*integral = j;
		*frac = x - j;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+		return;
+	}
+	// Check if `x` can even have a fractional part...
+	if ( exp > 51 ) {
+		// `x` is integral:
+		*frac = x;
+		*integral = 0.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
		*integral = x;
		*frac = 0.0;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+		return;
+	}
+	i = ALL_ONES >> ( exp - 20 );
+
+	// Determine if `x` is integral by checking for less significant significand bits which cannot be exponentiated away...
+	if ( ( low & i ) == 0 ) {
+		*frac = x;
+		*integral = 0.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
		*integral = x;
		*frac = 0.0;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+		return;
+	}
+	// Turn off all the bits which cannot be exponentiated away:
+	low &= ( ~i );
+
+	// Generate the integral part:
+	j = i;
+	stdlib_base_float64_from_words( high, low, &j );
+	*frac = j;
+	*integral = x - j;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
	stdlib_base_float64_from_words( high, low, &j );
	
	// The fractional part is whatever is leftover:
	*integral = j;
	*frac = x - j;
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Ahh, tests fail, @gunjjoshi can you please iterate over implementation again and see where the flaw is, something is off somewhere. Try it, else ping us, will take a close look.

```console
....
      ✖ returns integral part
      ✖ returns fractional part


  total:     10042
  passing:   6036
  failing:   4006
  duration:  391ms
```

**Comment by gunjjoshi** [comment]
Hey @Pranavchiku, all tests pass now. There was an issue after the last commit. Also, I've made changes as suggested. Thanks for pointing them out !

File: lib/node_modules/@stdlib/math/base/special/modf/lib/native.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+/**
+* Decomposes a double-precision floating-point number into integral and fractional parts, each having the same type and sign as the input value.
+*
+* @param {number} x - input value
+* @returns {Array<number>} output array
+*
+* @example
+* var parts = modf( 3.14 );
+* // returns [ 3.0, 0.14000000000000012 ]
+*
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/lib/native.js]:
```suggestion
```

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/modf/lib/native.js]: Please do not commit it, I will commit entire batch

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+	int32_t exp;
+	int32_t i;
+	double j;
+
+	// Special cases...
+	if ( x < 1.0 ) {
+		if ( x < 0.0 ) {
+			stdlib_base_modf( -x, integral, frac );
+			*frac *= -1.0;
+			*integral *= -1.0;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
			*integral *= -1.0;
			*frac *= -1.0;
```

This makes no difference but shows incorrect interpretation and creates confusion, better fix it.

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+			return;
+		}
+		if ( x == 0.0 ) { // [ +-0, +-0 ]
+			*frac = x;
+			*integral = x;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
			*integral = x;
			*frac = x;
```

File: lib/node_modules/@stdlib/math/base/special/modf/src/main.c [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+			return;
+		}
+		*integral = 0.0;
+		*frac = x;
+		return;
+	}
+	if ( stdlib_base_is_nan( x ) ) {
+		*frac = 0.0 / 0.0; // NaN
+		*integral = 0.0 / 0.0; // NaN
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/modf/src/main.c]:
```suggestion
		*integral = 0.0 / 0.0; // NaN
		*frac = 0.0 / 0.0; // NaN
```

File: lib/node_modules/@stdlib/math/base/special/modf/lib/native.js (hunk 15 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/modf/lib/native.js]: Hey @Pranavchiku, just a quick doubt, why are we not including an example in `native.js` for this one, unlike others ?

**Review by Pranavchiku — APPROVED** [review]
LGTM. Thanks @gunjjoshi !

File: lib/node_modules/@stdlib/math/base/special/modf/lib/native.js (hunk 15 above)
Comment (Pranavchiku) [reply lib/node_modules/@stdlib/math/base/special/modf/lib/native.js]: We have it, no?

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/modf/lib/native.js]: Oh, sorry, I mislooked it. Thanks for pointing !

**Pull request merged**